pre_commit/repository.py

```diff
@@ -64,11 +64,18 @@ def languages(self):
 
     @cached_property
     def additional_dependencies(self):
-        dep_dict = defaultdict(lambda: defaultdict(set))
+        dep_dict = defaultdict(lambda: defaultdict(list))
         for _, hook in self.hooks:
-            dep_dict[hook['language']][hook['language_version']].update(
-                hook.get('additional_dependencies', []),
-            )
+            for dep in hook.get('additional_dependencies', []):
+                if dep not in \
+                        dep_dict[hook['language']][hook['language_version']]:
+                    dep_dict[hook['language']][hook['language_version']].\
+                        append(dep)
+                else:
+                    logger.warning(
+                        'Additional dependency {} is listed twice for hook '
+                        '{}'.format(dep, hook['name'])
+                    )
```

Comment:
Maybe we can keep the code change smaller by making something like this:

```python
class _UniqueList(list):
    def __init__(self):
        self._set = set()

    def update(self, obj):
        for item in obj:
            if item not in self._set:
                self.append(item)
```

And then the diff is just

```diff
-        dep_dict = defaultdict(lambda: defaultdict(set))
+        dep_dict = defaultdict(lambda: defaultdict(_UniqueList))
```